dev-docs/RFCs/v6.1/transform-webgl1-support.md

```diff
@@ -0,0 +1,59 @@
+
+* **Author**: Ravi Akkenapally
+* **Date**: Aug, 2018
+* **Status**: **Review**
+
+Notes:
+* PR: TODO
+
+
+## Overview
+
+With the introduction of `Transform` class, luma.gl provided an easy to use API around WebGL2's complicated TransformFeedback API and also the ability to manage/create several other WebGL resources internally. This allows applications to offload complex operation to GPU and run in parallel while making CPU available for other tasks. But this class is limited to WebGL2 context and cannot be used with WebGL1 context. In this RFC we propose an implementation to support this class for WebGL1 context.
+
+## Main idea
+
+We should be able to setup a render pipeline (strictly using WebGL1 features), such that input parameters are provided as attributes to the VertexShader, inside VertexShader user provided operations are performed, and result is passed to FragmentShader, which writes the result to a pixel as fragment color.
+
+## Implementation details
+
+### Render target
+
+Size of the target should be such that it has one pixel per each expected output. I.e. if user wants to perform an operation on two float array of size 10K , then render target should have at least 10K pixels. Type of render target :
+
+#### Option#1 Framebuffer
+
+We can create a texture and use it as color attachment to a `Framebuffer` object. There might be limitation on some of the browsers using FLOAT texture as render target, for example Safari doesn't support `WEBGL_color_buffer_float` hence a float texture can't be used as color attachment. We might be able to overcome this issue using recently added [pack](https://github.com/uber/luma.gl/blob/master/modules/imageprocessing/src/modules/pack.md) shader module of luma.gl.
+
+#### Option#2 System color buffer
+
+In this option, instead of using Framebuffer object we can use default render color buffer. This has no limitations regarding float type, but can't be used with other graphics applications that render to system provided color buffer. But can be used with applications that do not render anything, like data processing and unit testing etc.
+
+We can start implementation with Option#1.
+
+Once the render target is setup, and data is rendered we can use `gl.readPixels` to read the data.
+
+### Vertex Processing
+
+Idea is to treat each input parameter as an attribute of a vertex and render all vertices as `gl.POINTS`. And perform required operations in Vertex Shader and pass the result into Fragment Shader as a varying. But to be able to place this result onto a single corresponding pixel, we need to also set `gl_Position` value in vertex shader. This position attribute can be internally calculated based on number of elements and set on vertex shader.
+
+### Fragment Processing
+
+Fragment shader will be just a pass through shader, it will move varying value into `gl_FragColor`.
+
+NOTE: Given this effort is to support WebGL1, all shaders must be in GLSL100 syntax.
+
+## Limitations
```

Comment:
Would like to see some code examples, maybe showing the following:

## Output

We can read from buffers (attributes) but we can't write to buffers, we can only write to textures, right. How would this look in an API?

## Chaining operations

To chain the first operation could take buffers or textures as input, additional operations need textures as input.

Sometimes textures can also be useful input in WebGL2... ?

## Readback to CPU

How do we read back data, when the output is sometimes in Buffers and sometimes in Textures, does transform offer a higher lever `getData()`

Reply:
> Output

Added a section on Input and Outputs, in phase#1 we can just do Buffer (requires sync) and phase#2 extend the API to support `Textures`?

> Sometimes textures can also be useful input in WebGL2... ?

Yes, possible, as updated in the RFC, we also have to address an issue having texture sample instructions in vertex shader, either force the API or insert these instructions under the hood.

> does transform offer a higher lever `getData()`

No ,  `Transform` supports `getBuffer` only, during the development we had higher level `getData()`, but decided discard it in favor not to duplicate `Buffer` API inside `Transform`. But now have a use case to support higher level `getData()`. 

I will update RFC with more code samples.